### ISSUE TITLE
Automated testing for Match options

### ISSUE THREAD

**Issue by jbellik** [syntax-prosody-ot/main#450]
See custom Match options testing documents: 
- https://github.com/syntax-prosody-ot/main/blob/master/test/aug20_customMatchSPTest_methods.pdf
- https://github.com/syntax-prosody-ot/main/blob/master/test/aug20_customMatchSP.html

**Comment by jbellik** [comment]
Create an html document that uses Mocha tests to check the values produced by various matchSP() and matchPS() option parameters. Basically, you should add it() lines with assert() statements for all the values that show up in the tableaux (tables). These statements should be grouped in a sensible way that recapitulates the organization of the table. That is, we could group the asserts for each column together, or (probably better) group the asserts for each row together.

* Refer to the html document linked above (aug20_customMatchSP.html) for trees to test the constraints on. Actually, it would make sense to a) rename this document to something like: customMatchSP_test_Mocha.html, and b) add the Mocha tests to it. Alternatively, you could break it up into a few different test files, if the current length is unmanageable. 

* Refer to the pdf document linked above (aug20_customMatchSPTest_methods.pdf) for expected return values (in linguist terms, violation counts) from the constraint functions matchSP and matchPS.

**Comment by maxTarlov** [comment]
matchSP sub-category sensitivity tested in #495 and #500, but automated testing with matchPS and visibility settings not yet implemented. I don't think we need to redo the entire test suit with matchPS, since all of the actual functionality is in matchSP.

**Proposal**: test each sub-category with matchPS, lexical and overtHead, but not each *combination* of sub-category (we already know that they work together, there is no good reason that should change when {requireOvertHead: true} is added to options).

**Comment by jbellik** [comment]
Testing MatchPS: Just redo tableaux 1 & 2
Or update old MatchPS testing files (such as file:///Users/jenny/Documents/GitHub/main/test/manual_testing/moreMatchOptions.html)

Testing lexical / overt head options: Jenny will think of trees to use for this.

**Comment by M130** [comment]
What should I test further for this one? I think Max and I did the matchTest.hmtl that test for the moreMatchOption.html. There is a comment about "Testing MatchPS: Just redo tableaux 1 & 2." Redo the Tableaux 1 & 2 in matchTest.html or moreMatchOption.html. Or create an automated test for moreMatchOption.html. Could you clarify a bit more about what to do with this one? Thanks!

**Comment by jbellik** [comment]
Yes, sorry this wasn't more specific. What I was thinking was to do the tests from the blocks that are labeled "Tableau 1" and "Tableau 2" in matchTest.html, but for MatchPS instead of MatchSP. That means:
1. Build a prosodic tree that has the same structure as the syntactic tree in Tableau 1 & 2, just with the categories relabeled. I.e., replace 'cp' with 'i', replace 'xp' with 'phi, and replace 'x0' with 'w'. Actually, come to think of it, I think that's what ptree1 in matchTest.html already is. Use this as ptree for all the tests of matchPS().
2. Build two syntactic trees that have the same structure as the matching and mismatching ptrees. I guess one can be the existing stree1, but you'll have to make the mismatching stree2.
2. Flip all the options that are called blahSyntax in the MatchSP test to be blahProsody in the MatchPS test, and flip all the options that are called blahProsody in the MatchSP test to be blahSyntax in the MatchPS test.
3. Expected violation counts do not need to change.

For example: for the 2nd item in the Tableau 1 test block, the matchSP test is:
let options = {"maxSyntax": true};
assert.equal(matchSP(stree1, ptree1, "xp", options), 0, message(stree1, ptree1, options));
assert.equal(matchSP(stree1, ptree2, "xp", options), 2, message(stree1, ptree2, options));

The matchPS test would be:
let options = {"maxProsody": true};
assert.equal(matchSP(stree1, ptree1, "xp", options), 0, message(stree1, ptree1, options));
assert.equal(matchSP(stree2, ptree1, "xp", options), 2, message(stree1, ptree2, options));

Let me know if you have further questions!

**Comment by jbellik** [comment]
handout: https://docs.google.com/document/d/1ZTzuzfmZfmQNXbXQ6Ax7T4I64EWE5s4nkehwLfDVr4U/edit#

**Comment by M130** [comment]
The second tableau seems to be working but the first tableau for stree2 is giving zero violations for everything.

**Comment by jbellik** [comment]
That's because you've set up the test to look for nodes with category "xp" in the prosodic tree. It should have category "phi". I think this is because I failed to fix the category in the copy-paste above, too -- sorry about that! I'll correct that comment now. I fixed the category in a couple of the tests in matchPSTest.js to check that that was the problem, too.

Not sure this matters but I noticed that the terminal ids are in a different order in stree1 than in ptree1: {{a {c}} {b} {d e f}} --> {(a (b (c d))) (e f)} Stree1 apparently has 'c' before 'b'.
